lib/sourcemapper.js

```diff
@@ -0,0 +1,253 @@
+ */
+function SourcemapperApi(infoMap) {
+  this.infoMap_ = infoMap;
+}
+
+/**
+ * Used to determine if the source file specified by the given path has
+ * a .map file and an output file associated with it.
+ * 
+ * If there is no such mapping, it could be because the input file is not
+ * the input to a transpilation process or it is the input to a transpilation
+ * process but its corresponding .map file was not given to the constructor
+ * of this mapper.
+ * 
+ * @param {string} inputPath The path to an input file that could
+ *  possibly be the input to a transpilation process.  The path should be 
+ *  relative to the process's current working directory.
+ */
+SourcemapperApi.prototype.hasMappingInfo = function(inputPath) {
+  return this.infoMap_.has(path.normalize(inputPath));
+};
+
+/**
+ * @param {string} inputPath The path to an input file that could possibly 
+ *  be the input to a transpilation process.  The path should be relative to 
+ *  the process's current working directory
+ * @param {number} The line number in the input file where the line number is
+ *   zero-based.
+ * @param {number} (Optional) The column number in the line of the file
+ *   specified where the column number is zero-based.
+ * @return {Object} The object returned has a "file" attribute for the
+ *   path of the output file associated with the given input file (where the 
+ *   path is relative to the process's current working directory),
+ *   a "line" attribute of the line number in the output file associated with
+ *   the given line number for the input file, and an optional "column" number
+ *   of the column number of the output file associated with the given file
+ *   and line information.
+ * 
+ *   If the given input file does not have mapping information associated
+ *   with it then null is returned.
+ */
+SourcemapperApi.prototype.mappingInfo = function(inputPath, lineNumber, colNumber) {
+  inputPath = path.normalize(inputPath);
+  if (!this.hasMappingInfo(inputPath)) {
+    return null;
+  }
+
+  var entry = this.infoMap_.get(inputPath);
+  var sourcePos = {
+    source: path.relative(path.dirname(entry.mapFile), inputPath),
+    line: lineNumber + 1, // the SourceMapConsumer expects the line number 
+                          // to be one-based but expects the column number 
+    column: colNumber     // to be zero-based
+  };
+
+  var consumer = entry.mapConsumer;
+  var allPos = consumer.allGeneratedPositionsFor(sourcePos);
+  /*
+  * Based on testing, it appears that the following code is needed to 
+  * properly get the correct mapping information.
+  * 
+  * In particular, the generatedPositionFor() alone doesn't appear to 
+  * give the correct mapping information.
+  */
+  var mappedPos = allPos && allPos.length > 0 ?
+    Array.prototype.reduce.call(allPos,
+      function(accumulator, value, index, arr) {
+          return value.line < accumulator.line ? value : accumulator;
+      }) : consumer.generatedPositionFor(sourcePos);
```

Comment:
In what case will `allGeneratedPositionsFor` return nothing but `generatedPositionFor` return something?